### PR TITLE
Add directives support for `createElementTemplate`

### PR DESCRIPTION
**Pull request by KingOfTac**
# Pull Request

## 📖 Description
This PR adds the ability to attach directives to the host element provided to `createElementTemplate` as well as consolidating the function's arguments into a single options object, which affects the `RenderInstruction.create` API.

After this change this is now possible:
```ts
class Model {
  @observable elementRef: HTMLElement;
  handleClick() {}
}

const template = RenderInstruction.createElementTemplate(
  "fast-button",
  {
    attributes: { "@click": x => x.handleClick() },
    directives: [ref("elementRef")], // applies to the "button" element being created
    content: html`<svg slot="start">...</svg> click me!`
  }
);
```

Prior to this change, it was not possible for the `Model` instance to access the `button` element it is bound to. In addition if the `Model` needed the children nodes of the host element, the template would need another `ViewTemplate` passed to as content with the `children` directive on it. Now the `children` directive can be directly applied to the host.

## 👩‍💻 Reviewer Notes
While everything tested locally just fine. I couldn't get the tests to detect when the property for a `ref` or `children` directive was updated. Any suggestions with writing tests for the new behavior would be appreciated.

Feedback on the new options type name and structure is also appreciated.

## 📑 Test Plan
Existing tests pass after this update. See above regarding new tests.

### General

- [X] I have included a change request file using `$ yarn change`
- [X] I have added tests for my changes.
- [X] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [X] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

### CODE REVIEW

**Review by EisenbergEffect — APPROVED** [review]

**Comment by EisenbergEffect** [comment]
Looks good. Just need to add a test. @kingoftac Can you share a bit more about the issue you had writing the test? I didn't quite understand.

**Comment by KingOfTac** [comment]
> Looks good. Just need to add a test. @kingoftac Can you share a bit more about the issue you had writing the test? I didn't quite understand.

I tried adding a test that would bind a ref to the element created by the method, the model would have the property as an observable, however in the test the ref would always be undefined, even when accessing it after calling await Updates.next()

Everything works as you would expect when using the code normally, it's just in the tests where it is having issues.

**Comment by chrisdholt** [comment]
> > Looks good. Just need to add a test. @KingOfTac Can you share a bit more about the issue you had writing the test? I didn't quite understand.
> 
> I tried adding a test that would bind a ref to the element created by the method, the model would have the property as an observable, however in the test the ref would always be undefined, even when accessing it after calling await Updates.next()
> 
> Everything works as you would expect when using the code normally, it's just in the tests where it is having issues.

If you want to push it up on the branch I'm happy to look and see what might be going on with you.

**Comment by KingOfTac** [comment]
@chrisdholt @EisenbergEffect Fixed my issue. It wasn't the test, the internal strings array and values array weren't lining up in some scenarios which resulted in valid HTML, however the directive wasn't getting bound.

File: packages/web-components/fast-element/src/templating/render.ts [hunk 1]
```diff
@@ -352,11 +368,27 @@ function createElementTemplate<TSource = any, TParent = any>(
...
     }
 
+    if (directives) {
+        markup[markup.length - 1] += " ";
+
+        for (let i = 0, ii = directives.length; i < ii; ++i) {
+            const directive = directives[i];
+
+            if (i === 0) {
+                markup.push(" ");
```
Comment (scomea) [inline packages/web-components/fast-element/src/templating/render.ts:381]: Could be written as -

`markup.push(i === 0  ? " " : "");`

**Review by scomea — APPROVED** [review]

**Comment by KingOfTac** [comment]
ping @chrisdholt. This is ready, I can update the branch for merging I just need to know the preferred merge strategy since I haven't done it in FAST before. It seems doing a merge commit has messed up the history in the past though.

**Comment by chrisdholt** [comment]
> ping @chrisdholt. This is ready, I can update the branch for merging I just need to know the preferred merge strategy since I haven't done it in FAST before. It seems doing a merge commit has messed up the history in the past though.

I've updated and I can merge - just as an FYI, we should have this covered (squash merge is only enabled): 
image

**Pull request merged**